### ISSUE TITLE
Skills: Unknown intent is not an option for a task when creating a manifest.

### ISSUE THREAD

**Issue by JonathanFingold** [microsoft/BotFramework-Composer#7470]
## Describe the bug

The unknown intent is the only logical task for my echo skill, as designed. I can't choose the unknown intent as the task the skill can perform.

## Version

Version:  2.0.0-nightly.240624.c8283fa
Electron: 8.2.4
Chrome: 80.0.3987.165
NodeJS: 12.13.0
V8: 8.0.426.27-electron.0

## Browser

- [x] Electron distribution
- [ ] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Edge

## OS

- [ ] macOS
- [x] Windows
- [ ] Ubuntu

## To Reproduce

1. Create a basic echo skill, once that does something based on any message that comes in.
2. Try to share the bot as a skill
3. Error: when you get to the **Select which tasks this skill can perform** dialog, the unknown intent trigger is not listed as an opion.

## Expected behavior

A generic message activity should also be an option for a task a skill can start with, and presumably a generic event or invoke, too.

## Screenshots

![image](https://user-images.githubusercontent.com/35047199/116485284-b2235c80-a83f-11eb-8cc7-146c776261ad.png)

## Additional context

Trying to figure out how to explain this wizard in the how to publish a skill article.

**Comment by luhan2017** [comment]
The unknown intent sounds like a by-default choice. I think user don't need to select it specifically, it is naturally supported. Either select or non-select it, the manifest will be generated for the same content. @darrenj @gabog, thoughts?

**Comment by JonathanFingold** [comment]
I will run with that assumption for now. The messaging in the UI is confusing on that front, but can be fixed later.

**Comment by JonathanFingold** [comment]
The default appears to be **event**, instead of **unknown intent**!

```json
  "activities": {
    "EchoSkill": {
      "type": "event",
      "name": "EchoSkill",
      "description": "Skill for the local skills how to."
    }
  }
```

**Comment by vivekkshankar** [comment]
@luhan2017 Is this behaviour by design?
I.e we're assuming there insn't a scenario where a skill would handle unknown intent but the root bot would and therefore we dont expose it as a choice. We should verify this.

if its not by design or if we default to it - it should be exposed in the UI

@darrenj @mareekuh Can you help with some context here?

cc @mewa1024

**Comment by JonathanFingold** [comment]
My 2 cents:
- Intents are sub types of message activities.
- Unknown intents are a sub type of all activities, effectively.
- If we don't allow the unknown intent to be included in the manifest, we don't allow a skill to handle a generic message without a recognizer in the consumer bot.

@gabog may be able to clarify.

**Comment by tdurnford** [comment]
@JonathanFingold When a skill receives a message activity it runs the dialogs recognizer to determine the intent. At that point, if the result is an unknown intent, the `OnUnknownIntent` handler is trigger exactly like how a known intent would trigger the `OnIntent` handler.

For the same reason `OnIntent` triggers aren't included in the generate skill manifest wizard, `OnUnknownIntent` triggers are not included. This is by design.

**Comment by JonathanFingold** [comment]
The skill defines three triggers: **Greeting** (ConversationUpdate), **Unknown intent**, and **stop** (Intent recognized). The skills manifest wizard lets me choose from **Greeting** and **stop**. When I chose none of them, the generated manifest included this:

```
  "activities": {
    "EchoSkill": {
      "type": "event",
      "name": "EchoSkill",
      "description": "Skill for the local skills how to."
    }
  }
```

**Comment by tdurnford** [comment]
@JonathanFingold After looking at the code closer, you are correct. The 'Select tasks' step should include the `OnUnknownIntent` handler; however, it will just add a message activity to the activities. This is what the activities property should look like for selecting all the triggers including OnUnknowIntent in your scenario: 

```
  "activities": {
     "message": {  <--- Message activity for both intents and unknown intents
      "type": "message"
    },
    "EchoSkill": {
      "type": "event",
      "name": "EchoSkill",
      "description": "Skill for the local skills how to."
    }
  }
```

Not sure why selecting none of them would result in the object having an event.

**Comment by darrenj** [comment]
I'm a bit confused about the real-world scenario underpinning this, happy to be educated :-)

If I write a Bot which is intended to be used a skill I would have intents configured using one of the available recognizers. i.e. I'll have dialogs trigged by intents.  The triggers for these intents will have LU based training data so the bot knows where to direct each utterance.

When we switch to a skill, the goal of the manifest is to give the parent bot the information it needs to know when to pass an utterance across to the Skill. Utterances the Skill cannot handle should therefore not be passed across and handled by the parent bot and it's unknown handler as needed.

Therefore, having two levels of unknown intent handling doesn't feel right as the skill can say hey I've got all these utterances I can't handle but I'm going to advertise that I can meaning I may "steal" them away from the parent bot. I would therefore say that we should not support adding the unknown intent handler as per the current design. Open to discussion though!

**Comment by tdurnford** [comment]
@darrenj You are correct, the manifest is intended to communicate types of activities the skill can handle. When the developer adds an `OnIntent` trigger, we add a message activity to the manifest to indicate the skill can accept and process message activities. Adding an intent would result in the following manifest:

```
{
  ...
  "activities":  {
    "message": {
      "type": "message"
  }
}
```

If we allow a developer to add an `OnUnknownIntent`, they would be indicating that the skill can accept and process message activities, so the manifest would be exactly the same. Adding the ability to add `OnUnknownIntent` does not add much value, unless the skills only intent trigger is the `OnUnknownIntent` which is an edge case.

I'm on the fence about whether or not we should add `OnUnknownIntent` to the `Select task` step. It's probably more of a design question than anything else.

**Comment by JonathanFingold** [comment]
@darrenj The problem is that we're mixing two aspects here, and it gets confusing:
1. The root bot can use the manifest information to build a language model for utterances that the skill can handle, and that is reflected in the OnIntent triggers in the skill.
2. However, the skill bot is expecting activities (not triggers). So a basic skill might just want a message or a conversation update activity to kick it off. We're forcing dev to design their skills in a specific way that is outside of the scope of the manifest schema.

This is probably not a ship blocker, but it does mean:
1. I have to explain this additional constraint to users to avoid confusion.
2. We should be aware that we're adding a design constraint to the skills feature in Composer.

**Comment by darrenj** [comment]
I've talked this through with a few to validate - I was right this is not a valid scenario. A skill should not advertise things it cannot call. A bot (or skill) should not wire up everything in say an unknown handler and should instead make use of triggers as that's the pattern for Composer first bots.

I don't believe this is a constraint and needs to be documented differently. However the copy on the UX does need changing and I believe is in flight @vivekkshankar ?

**Comment by mareekuh** [comment]
@vivekkshankar Did I already include this content or is it additional. LMK asap so I can add it - guys are implementing copy as we speak.
